core/utils/utils.go

```diff
@@ -520,3 +522,111 @@ func Uint256ToHex(n *big.Int) (string, error) {
 func DecimalFromBigInt(i *big.Int, precision int32) decimal.Decimal {
 	return decimal.NewFromBigInt(i, -precision)
 }
+
+// FIFO queue that discards older items when it reaches its capacity.
+type BoundedQueue struct {
+	capacity uint
+	items    []interface{}
+	mu       *sync.RWMutex
+}
+
+func NewBoundedQueue(capacity uint) *BoundedQueue {
+	return &BoundedQueue{
+		capacity: capacity,
+		mu:       &sync.RWMutex{},
+	}
+}
+
+func (q *BoundedQueue) Add(x interface{}) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.items = append(q.items, x)
+	if uint(len(q.items)) > q.capacity {
+		excess := uint(len(q.items)) - q.capacity
+		q.items = q.items[excess:]
```

Comment:
Could this excess ever be more than 1? Do we need this if block? Maybe slightly less efficient if we trim every time, even when excess equals zero.

Reply:
Currently true for FM, but I want to push us to rely on more reusable, generalizable components that we can apply to other parts of the codebase.  Less testing and better reliability that way.  I've gotten a lot of mileage out of the ones we already have, most recently `BackoffSleeper`.

Re: performance, at present, we don't receive more than a few dozen logs per second, so eliminating this seems unlikely to make an impact (a comparison and two arithmetic ops).  However, with components like these, we should absolutely do regular performance profiling with `pprof` to ensure that they don't ever become bottlenecks.